File: 0152-maximum-product-subarray/0152-maximum-product-subarray.go
func maxProduct(nums []int) int {
    res := nums[0];
    minEnding := nums[0];
    maxEnding := nums[0];
    for i := 1; i < len(nums); i++ {
        curr := nums[i];
        prevMin := minEnding;
        prevMax := maxEnding;
        maxEnding = max(curr, curr * prevMin, curr * prevMax);
        minEnding = min(curr, curr * prevMin, curr * prevMax);
        res = max(res, maxEnding);
    }
    return res;
}

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int res = nums[0];
        int minEnding = nums[0];
        int maxEnding = nums[0];
        for (int i = 1; i < nums.size(); i++) {
            int curr = nums[i];
            int prevMin = minEnding;
            int prevMax = maxEnding;
            maxEnding = max({curr, curr * prevMax, curr * prevMin});
            minEnding = min({curr, curr * prevMax, curr * prevMin});
            res = max(res, maxEnding);
        }
        return res;
    }
};